odxtools/diaglayer.py

```diff
@@ -636,9 +643,73 @@ def get_comparam(
 
         return cps[0]
 
-    def get_can_receive_id(self, protocol_name: Optional[str] = None) -> Optional[int]:
+    def get_can_mtu(self, protocol: Optional[Union[str, "DiagLayer"]] = None) -> Optional[int]:
+        """Return the maximum size of CAN frames that can be
+        transmitted in bytes.
+
+        For classic CAN busses, this is basically always 8. CAN-FD can
+        send up to 64 bytes per frame.
+        """
+        com_param = self.get_comparam("CP_CANFDTxMaxDataLength", protocol=protocol)
+        if com_param is None:
+            rx_id = self.get_can_receive_id(protocol=protocol)
+            if rx_id is not None:
+                # bus is CAN. We also use 8 bytes for CAN-FD if this
+                # parameter was not specified.
+                return 8
+
+            # bus is not CAN
+            return None
+
+        val = com_param.value
+        if not isinstance(val, str):
+            return None
+
+        m = re.search("TX_DL *= *([0-9]*)", val)
+        if m:
+            return int(m.group(1))
+
+        # unexpected format of parameter value
+        return 8
+
+    def use_can_fd(self, protocol: Optional[Union[str, "DiagLayer"]] = None) -> bool:
+        return (self.get_can_mtu(protocol=protocol) or 0) > 8
+
+    def get_can_baudrate(self, protocol: Optional[Union[str, "DiagLayer"]] = None) -> Optional[int]:
+        """Baudrate of the CAN bus which is used by the ECU"""
+        com_param = self.get_comparam("CP_Baudrate", protocol=protocol)
+        if com_param is None:
+            return None
+
+        val = com_param.value
+        if not isinstance(val, str):
+            return None
+
+        return int(val)
+
+    def get_can_fd_baudrate(self,
```

Comment:
Mostly duplicates `get_can_baudrate`, maybe add a `_get_baudrate(protocol, param_name)` helper?

Reply:
this is the signalling rate for the payload portion of CAN-FD frames. You still need the CAN baudrate for the rest of the frame. (i.e., CAN-FD busses exhibit two bitrates. this is to enable to continue to use non-FD devices on FD busses...)

Reply:
I am proposing this
```python
def get_can_baudrate(self, ....):
    self._get_baudrate(self, "CP_Baudrate", ...)

def get_can_fd_baudrate(self, ....):
    self._get_baudrate(self, "CP_CANFDBaudrate", ...)
```

Reply:
what would be the difference from the user's POV? maybe it is better to call this method `get_can_fd_databaudrate()`, though?

Reply:
(also, I do not believe that this would make things easier to understand or even save a significant number of lines of code...)

Reply:
a more generic approach would be
```python
def get_simple_comparam_value(self, comparam_name, conversion_fn)
```
so that most of these convenience function would reduce to
```python
def get_can_baudrate(self):
    return self.get_simple_comparam_value("CP_Baudrate", int)
```

(not sure if this a good idea either...)

Reply:
You can keep the `get_simple_comparam_value` private, that way you can change it if you want later

Reply:
I tried implementing this approach: It didn't work because I haven't found a good way to make it play nicely with type checking: The issue is that the return type of that function would be the return type of the callable passed as the second parameter...